internal/store/common.go

```diff
@@ -152,6 +156,66 @@ func GetNonNilFieldsFromResource(resource model.Resource) []string {
 	return ret
 }
 
+func matchExpressionQueryAndArgs(matchExpression api.MatchExpression, colName string) (string, []any) {
+	params := strings.Join(lo.Map(lo.FromPtr(matchExpression.Values), func(_ string, _ int) string { return "?" }), ",")
+	args := lo.Map(lo.FromPtr(matchExpression.Values), func(val string, _ int) any { return matchExpression.Key + "=" + val })
+	return fmt.Sprintf("%s && ARRAY[%s]", colName, params), args
+}
+
+func matchInQuery(query *gorm.DB, matchExpression api.MatchExpression, colName string) *gorm.DB {
+	whereStr, args := matchExpressionQueryAndArgs(matchExpression, colName)
+	return query.Where(whereStr, args...)
+}
+
+func matchNotInQuery(query *gorm.DB, matchExpression api.MatchExpression, colName string) *gorm.DB {
+	whereStr, args := matchExpressionQueryAndArgs(matchExpression, colName)
+	return query.Not(whereStr, args...)
+}
+
+func existsQuery(colName string) string {
+	return fmt.Sprintf("exists (select 1 from unnest(%s) as element where element like ?)", colName)
+}
+
+func matchExistsQuery(query *gorm.DB, matchExpression api.MatchExpression, colName string) *gorm.DB {
+	return query.Where(existsQuery(colName), matchExpression.Key+"=%")
+}
+
+func matchDoesNotExistQuery(query *gorm.DB, matchExpression api.MatchExpression, colName string) *gorm.DB {
+	return query.Not(existsQuery(colName), matchExpression.Key+"=%")
+}
+
+func matchExpressionQuery(query *gorm.DB, matchExpression api.MatchExpression, colName string) *gorm.DB {
+	switch matchExpression.Operator {
+	case api.In:
+		return matchInQuery(query, matchExpression, colName)
+	case api.NotIn:
+		return matchNotInQuery(query, matchExpression, colName)
+	case api.Exists:
+		return matchExistsQuery(query, matchExpression, colName)
+	case api.DoesNotExist:
+		return matchDoesNotExistQuery(query, matchExpression, colName)
+	default:
+		// TODO: Return error for this case and handle it
+		return query
+	}
+}
+
+func matchExpressionsQuery(query *gorm.DB, matchExpressions api.MatchExpressions, colName string) *gorm.DB {
+	for _, me := range matchExpressions {
+		query = matchExpressionQuery(query, me, colName)
```

Comment:
Would it be better to apply matchExpressionQuery within this loop and eliminate the standalone matchExpressionQuery function? We can conceptualize that even a single expression is part of a matching set

Reply:
Why ?  Is it better ?